File: javademo/src/main/java/kr/sean/javademo/MainActivity.java
package kr.sean.javademo;

import android.os.Bundle;
import android.os.HandlerThread;
import android.view.KeyEvent;
import android.view.View;
import android.widget.Button;

import androidx.appcompat.app.AppCompatActivity;


import com.github.anrwatchdog.ANRWatchDog;

import org.sean.kim.jag.C;
import org.sean.kim.jag.Jag;
import org.sean.kim.jag.JagDefaultWatchDog;
import org.sean.kim.jag.JagResult;
import org.sean.kim.jag.util.Logger;
import org.sean.kim.jag.util.Blocking;

import java.io.IOException;

public class MainActivity extends AppCompatActivity {
    private Logger logger = new Logger(C.TAG, "Demo");
    private Button button1;
    private Button button2;
    private Jag jag;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        HandlerThread th = new HandlerThread("JagWatchDog");
        th.start();
        jag = new Jag(this);
        //jag = new Jag(this, new ANRWatchDog(4500));
        //jag = new Jag(this, new JagDefaultWatchDog(4500));
        button1 = findViewById(R.id.button1);
        button1.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                test0();
            }
        });
        button2 = findViewById(R.id.button2);
        button2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                test1();
            }
        });
        findViewById(R.id.button3).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                test2();
            }
        });
    }

    @Override
    protected void onDestroy() {
        jag.release();
        super.onDestroy();
    }

    @Override
    public boolean onKeyUp(int keyCode, KeyEvent event) {
        return super.onKeyUp(keyCode, event);
    }

    private void test0() {
        JagResult<Boolean> result = jag.request(()->Blocking.fileBlockingLongExecution(this));
        if (result.get(false)) button1.setText("Success");
        else button1.setText("Failed");
    }

    private void test1() {
        JagResult<Boolean> result = jag.request(()->Blocking.fileBlockingLongExecution(this));
        result.onComplete((Boolean success)->{
            logger.i("test1: %s", success ? "Success" : "Failed");
            button2.setText(success ? "Success" : "Failed");
        });
        logger.i("waiting...");
    }

    private void test2() {
        try {
            Blocking.fileBlockingLongExecution(this);
        } catch (IOException e) {
            throw new RuntimeException(e);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

}

File: demo/src/main/java/org/sean/kim/jag/demo/MainActivity.kt
package org.sean.kim.jag.demo

import android.os.Bundle
import android.view.KeyEvent
import android.widget.Button
import androidx.appcompat.app.AppCompatActivity
import org.sean.kim.jag.C
import org.sean.kim.jag.Jag
import org.sean.kim.jag.util.Blocking
import org.sean.kim.jag.util.Logger

class MainActivity : AppCompatActivity() {
    private val logger: Logger = Logger(C.TAG, "Demo")
    private lateinit var jag: Jag
    private lateinit var button1: Button

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        jag = Jag(this)
        button1 = this.findViewById(R.id.button1)
        button1.setOnClickListener {
            test0()
        }
    }

    override fun onKeyUp(keyCode: Int, event: KeyEvent?): Boolean {
        when (keyCode) {
            KeyEvent.KEYCODE_BACK -> {
                finish()
                return true
            }
            KeyEvent.KEYCODE_0 -> {
                test0()
                return true
            }
        }
        return super.onKeyUp(keyCode, event)
    }

    private fun test0() {
        jag.request { Blocking.fileBlockingLongExecution(this) }.let {
            it.get(false).let {
                logger.d("test0", "result: $it")
                if (it) button1.text = "Success"
                else button1.text = "Failed"
            }
        }
    }

}